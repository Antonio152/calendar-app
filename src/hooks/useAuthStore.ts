import { useSelector } from 'react-redux'
import { RootState, useAppDispatch } from '../store/store'
import { calendarAPI } from '../api/calendarAPI'
import {
  clearErrorMessage,
  onChecking,
  onLogin,
  onLogout
} from '../store/auth/authSlice'
import { IDataResult, IRenewToken } from './types/useAuthStoreTypes'
import { onLogoutCalendar } from '../store/calendar/calendarSlice'

export const useAuthStore = () => {
  const dispatch = useAppDispatch()
  const { status, user, errorMessage } = useSelector(
    (state: RootState) => state.auth
  )

  const startRegisterUser = async (
    name: string,
    email: string,
    password: string
  ) => {
    dispatch(onChecking())
    try {
      const { data }: IDataResult = await calendarAPI.post('/auth/new', {
        name,
        email,
        password
      })
      localStorage.setItem('token', data.token)
      localStorage.setItem('token-init-date', new Date().getTime().toString())
      dispatch(onLogin({ uid: data.uid, name: data.name }))
    } catch (error:any) {
      dispatch(
        onLogout(error.response.data.msg || 'Error al crear el usuario, revisa que los datos ingresados sean correctos')
      )
      setTimeout(() => {
        dispatch(clearErrorMessage())
      }, 30)
    }
  }

  const startLogin = async (email: string, password: string) => {
    dispatch(onChecking())
    try {
      const { data }: IDataResult = await calendarAPI.post('/auth', {
        email,
        password
      })
      localStorage.setItem('token', data.token)
      localStorage.setItem('token-init-date', new Date().getTime().toString())
      dispatch(onLogin({ uid: data.uid, name: data.name }))
    } catch (error) {
      console.log(error)
      dispatch(onLogout('Credenciales incorrectas'))
      setTimeout(() => {
        dispatch(clearErrorMessage())
      }, 30)
    }
  }

  const startLogout = () => {
    localStorage.clear()
    dispatch(onLogout(''))
    dispatch(onLogoutCalendar())
  }
  const checkAuthToken = async () => {
    const token = localStorage.getItem('token')
    if (!token) {
      dispatch(onLogout(''))
    }
    try {
      const { data }:IRenewToken = await calendarAPI.get('/auth/renew')
      localStorage.setItem('token', data.token)
      localStorage.setItem('token-init-date', new Date().getTime().toString())
      dispatch(onLogin({ uid: data.uid, name: data.name }))
    } catch (error) {
      localStorage.clear()
      dispatch(onLogout(''))
    }
  }
  return {
    // * Propieties
    status,
    user,
    errorMessage,
    // * Functions
    startRegisterUser,
    startLogin,
    startLogout,
    checkAuthToken
  }
}
