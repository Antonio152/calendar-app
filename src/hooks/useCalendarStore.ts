import { useSelector } from 'react-redux'
import { RootState, useAppDispatch } from '../store/store'
import { IBigCalendarEvent, IGetEvents } from '../calendar/types/CalendarTypes'
import { onAddNewEvent, onDeleteEvent, onLoadEvents, onSetActiveNote, onUpdateEvent } from '../store/calendar/calendarSlice'
import { calendarAPI } from '../api/calendarAPI'
import { IUseCalendarStore } from './types/useCalendarStoreTypes'
import { convertCalendarEventsToDateEvents } from '../helpers/convertCalendarEventsToDateEvents'
import Swal from 'sweetalert2'

export const useCalendarStore = () => {
  const dispatch = useAppDispatch()
  const { user } = useSelector((state: RootState) => state.auth)
  const { events, activeEvent } = useSelector((state: RootState) => state.calendar)

  const onSetActiveNoteH = (event: IBigCalendarEvent) => {
    dispatch(onSetActiveNote(event))
  }

  const onSaveCalendarEvent = async (eventCalendar: IBigCalendarEvent) => {
    try {
      if (eventCalendar.id) {
        // Update event
        await calendarAPI.put(`/events/${eventCalendar.id}`, eventCalendar)
        dispatch(onUpdateEvent({ ...eventCalendar, user: { _id: user.uid, ...user } }))
        return
      }
      // Add event
      const { data }:IUseCalendarStore = await calendarAPI.post('/events', eventCalendar)
      dispatch(onAddNewEvent({ ...eventCalendar, id: data.newEvent.id, user: { _id: user.uid, ...user } }))
    } catch (error) {
      console.log(error)
      Swal.fire('Error', 'Error al guardar el evento', 'error')
    }
  }

  const onDeleteEventH = async (event: IBigCalendarEvent) => {
    try {
      await calendarAPI.delete(`/events/${event.id}`)
      dispatch(onDeleteEvent(event))
    } catch (error) {
      console.log(error)
      Swal.fire('Error', 'Error al eliminar el evento', 'error')
    }
  }

  const startLoadingEvents = async () => {
    try {
      const { data }:IGetEvents = await calendarAPI.get('/events')
      const events = convertCalendarEventsToDateEvents(data.events)
      dispatch(onLoadEvents(events))
    } catch (error) {
      console.log(error)
      Swal.fire('Error', 'Error al cargar los eventos', 'error')
    }
  }
  return {
    events,
    activeEvent,
    onSetActiveNoteH,
    onSaveCalendarEvent,
    onDeleteEventH,
    startLoadingEvents
  }
}
