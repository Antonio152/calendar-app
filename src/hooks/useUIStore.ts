import { useSelector } from 'react-redux'
import { RootState, useAppDispatch } from '../store/store'
import { onCloseDateModal, onOpenDateModal } from '../store/ui/uiSlice'

export const useUIStore = () => {
  const dispatch = useAppDispatch()
  const { isDateModalOpen } = useSelector((state: RootState) => state.ui)

  const onOpenDateModalH = () => dispatch(onOpenDateModal())
  const onCloseDateModalH = () => dispatch(onCloseDateModal())

  return {
    isDateModalOpen,
    onOpenDateModalH,
    onCloseDateModalH
  }
}
