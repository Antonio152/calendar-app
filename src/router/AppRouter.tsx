import React, { useEffect } from 'react'
import { Routes, Route, Navigate } from 'react-router-dom'
import { LoginPage } from '../auth/pages/Login/LoginPage'
import { CalendarPage } from '../calendar/pages/CalendarPage'
import { useAuthStore } from '../hooks/useAuthStore'

export const AppRouter = () => {
  const { status, checkAuthToken } = useAuthStore()

  useEffect(() => {
    checkAuthToken()
  }, [])

  if (status === 'checking') {
    return <h1>Cargando...</h1>
  }

  return (
    <Routes>
        {
            (status !== 'authenticated')
              ? <>
                  <Route path="/auth/*" element={<LoginPage/>}/>
                  <Route path="/*" element={<Navigate to='/auth/login'/>} />
                </>
              : <>
                  <Route path="/" element={<CalendarPage/>}/>
                  <Route path="/*" element={<Navigate to='/'/>} />
                </>
        }

    </Routes>
  )
}
