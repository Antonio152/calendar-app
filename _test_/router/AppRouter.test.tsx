import React from 'react'
import { MemoryRouter } from 'react-router-dom'
import { render } from '@testing-library/react'
import { AppRouter } from '../../src/router/AppRouter'
import { useAuthStore } from '../../src/hooks/useAuthStore'
jest.mock('../../src/calendar/pages/CalendarPage', () => ({
  CalendarPage: jest.fn(() => <div>Calendar Page</div>)
}))

jest.mock('../../src/hooks/useAuthStore')
describe('testing in AppRouter', () => {
  const checkMock = jest.fn()
  beforeEach(() => jest.clearAllMocks())
  test('should return the loading scrren ', () => {
    (useAuthStore as jest.Mock).mockReturnValue({ status: 'checking', checkAuthToken: checkMock })
    const { getByText } = render(<AppRouter />)
    expect(getByText('Cargando...')).toBeTruthy()
    expect(checkMock).toHaveBeenCalledTimes(1)
  })
  test('should render the login page', () => {
    (useAuthStore as jest.Mock).mockReturnValue({ status: 'not-authenticated', checkAuthToken: checkMock })
    const { getByText, container } = render(<MemoryRouter>
        <AppRouter />
    </MemoryRouter>)
    expect(getByText('Inicia sesiÃ³n')).toBeTruthy()
    expect(container).toMatchSnapshot()
  })
  test('should render the calendar page', () => {
    (useAuthStore as jest.Mock).mockReturnValue({ status: 'authenticated', checkAuthToken: checkMock })
    const { getByText } = render(<MemoryRouter>
        <AppRouter />
    </MemoryRouter>)
    expect(getByText('Calendar Page')).toBeTruthy()
  })
})
