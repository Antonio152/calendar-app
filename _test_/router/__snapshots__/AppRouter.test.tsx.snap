// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing in AppRouter should render the login page 1`] = `
<div
  aria-hidden="true"
>
  <div
    class="col-9 m-auto d-flex bLogin centerLogin bg-white"
  >
    <div
      class="col-lg-5 col-md-5 d-flex justify-content-center noShowImageLogin"
    >
      <img
        alt=""
        class="imgContain"
        height="100%"
        src="/images/login-inicio.gif"
        width="100%"
      />
    </div>
    <div
      class="col-lg-7 col-md-7 col-sm-12 m-auto p-5"
    >
      <div>
        <form>
          <h2
            class="mb-3 text-center"
          >
            Inicia sesión
          </h2>
          <div
            class="form-floating mb-3"
          >
            <input
              class="form-control"
              id="email"
              name="email"
              placeholder="Email"
              type="email"
              value=""
            />
            <label
              for="email"
            >
              Email
            </label>
          </div>
          <div
            class="form-floating mb-3"
          >
            <input
              class="form-control"
              id="password"
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
            <label
              for="password"
            >
              Password
            </label>
          </div>
          <div
            class="text-end mt-3 mb-3"
          >
            <a
              href="#"
            >
              ¿Olvidaste tu contraseña?
            </a>
          </div>
          <div
            class="d-grid gap-2"
          >
            <button
              class="btn btn-primary"
              type="submit"
            >
              Iniciar sesión
            </button>
          </div>
          <div
            class="text-center mt-3"
          >
            <span>
              Tambien puedes iniciar sesión con:
            </span>
            <div
              class="d-flex gap-4 mt-3 justify-content-center"
            >
              <button
                class="btn btn-link btn-floating mx-1"
                type="button"
              >
                <i
                  class="fa-brands fa-google"
                />
              </button>
              <button
                class="btn btn-link btn-floating mx-1"
                type="button"
              >
                <i
                  class="fa-brands fa-facebook"
                />
              </button>
              <button
                class="btn btn-link btn-floating mx-1"
                type="button"
              >
                <i
                  class="fa-brands fa-twitter"
                />
              </button>
            </div>
          </div>
          <div
            class="text-center mt-4 "
          >
            <span>
              ¿No eres miembro? 
               
              <span
                class="registerTxt"
              >
                Registrate
              </span>
            </span>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
